### PR TITLE
V3/remove this throw call transaction h

### PR DESCRIPTION
**Pull request by gberkes**
Introducing the use of assertions to address throw; calls that lack try-catch blocks. Upon examining the caller code that utilized methods containing the questioned throw; calls, it became clear that, in the current state of development, there are no scenarios where execution could reach these throw; calls. However, we cannot guarantee this for future development. For instance, if someone attempts to use getCurrentMarker() without first verifying isInsideAMarker(), ModSecurity would encounter the throw; and terminate. The issue with the other throw; call is similar in that it is, fortunately, unreachable at the moment. However, it differs because this throw; is intended to handle a case that has not yet been developed.

Fortunately, I found an article titled "Effective Use of Assertions in C++" by Mike A. Martin in (ACM SIGPLAN Language Tips, page 3), which offers a neat way to handle such cases, specifically regarding argument validation and unreachable code. [Link to the article](https://dl.acm.org/doi/pdf/10.1145/240964.240969).

Following the guidance from this article, I addressed the issues and also included modifications to enrich assert error messages. Furthermore, I updated configure.ac to maintain the usual build procedure and modified README.md to introduce the new configure flag.

### CODE REVIEW

File: headers/modsecurity/transaction.h [hunk 1]
```diff
@@ -307,11 +308,8 @@ class TransactionSecMarkerManagement {
     }
 
     std::shared_ptr<std::string> getCurrentMarker() const {
-        if (m_marker) {
-            return m_marker;
-        } else {
-            throw;
```
Comment (airween) [inline headers/modsecurity/transaction.h:313]: SonarCloud report is [here](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfJimrsSpWCKX0vrq)

File: src/rule_with_actions.cc [hunk 2]
```diff
@@ -121,10 +122,8 @@ RuleWithActions::RuleWithActions(
                     m_actionsRuntimePos.push_back(a);
                 }
             } else {
+                assert(false && "The handling of RunTimeBeforeMatchAttemptKind has not been implemented yet.");
                 delete a;
-                std::cout << "General failure, action: " << a->m_name;
-                std::cout << " has an unknown type." << std::endl;
-                throw;
```
Comment (airween) [inline src/rule_with_actions.cc:127]: SonarCloud report is [here](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfKAnrsSpWCKX0waI)

**Comment by airween** [comment]
Just for the record: this PR fixes two SonarCloud issues in files:

* [headers/modsecurity/transaction.h](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfJimrsSpWCKX0vrq)
* [src/rule_with_actions.cc](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfKAnrsSpWCKX0waI)

I also added these references as conversation.

File: README.md [hunk 3]
```diff
@@ -231,7 +231,7 @@ CFLAGS to disable the compilation optimization parameters:
 ```shell
 $ export CFLAGS="-g -O0"
 $ ./build.sh
-$ ./configure
+$ ./configure --enable-assertions=yes
```
Comment (airween) [inline README.md:240]: Please explain the behavior of this new option. First it's enough to place that in a single comment.

Later we should consider to create a separated documentation / extend the existing one with all options.

**Review by airween — CHANGES_REQUESTED** [review]

File: headers/modsecurity/transaction.h [hunk 4]
```diff
@@ -323,7 +321,7 @@ class TransactionSecMarkerManagement {
     }
 
  private:
-    std::shared_ptr<std::string> m_marker;
+    std::shared_ptr<std::string> m_marker = nullptr;
```
Comment (MirkoDziadzka) [inline headers/modsecurity/transaction.h]: I think C++ gurantees this. `std::shared_ptr` is a class and therefore has a constructor and initialize itself.

File: src/rule_with_actions.cc (hunk 2 above)
Comment (MirkoDziadzka) [inline src/rule_with_actions.cc]: What we are doing here is a switch over `a->action_kind` withe the 2 cases `Action::ConfigurationKind` and `Action::RunTimeOnlyIfMatchKind`.

This `else` would be the default case of the switch, right.

Would it be better to replace the if/else with a switch and ask the compiler of our switch case is complete (e.g. all cases are implemented)

----

independent from this remark: Is the else case a "can never happen" or is it a "user configured invalid modsec code". In the second case, we change the semantic which could be a problem.

File: README.md (hunk 3 above)
Comment (gberkes) [reply README.md:240]: Done: https://github.com/owasp-modsecurity/ModSecurity/pull/3104/commits/2c8684eeb0e99b25a4d488d802910cba927153bb

File: headers/modsecurity/transaction.h (hunk 4 above)
Comment (gberkes) [reply headers/modsecurity/transaction.h]: Yes. You're right. Fixed: https://github.com/owasp-modsecurity/ModSecurity/pull/3104/commits/cee3965f689ea52004d0e8b160054e227e6c5593

File: src/rule_with_actions.cc (hunk 2 above)
Comment (gberkes) [reply src/rule_with_actions.cc]: The refactoring of the if-then chain to a switch statement is complete. We reviewed the code together with @airween  and couldn't find a way to force execution into the else/default case.

**Comment by airween** [comment]
Please pull the modifications from #3134, and apply this PR. Or you can resolve the conflicts [here](https://github.com/owasp-modsecurity/ModSecurity/pull/3104/conflicts).

**Comment by huberts90** [comment]
@gberkes
Could you please address requested change to complete this issue?

**Pull request closed without merging**